### PR TITLE
switch hook-me from inlets to quic-reverse-proxy

### PR DESCRIPTION
**Pull request by deitch**
Signed-off-by: Avi Deitcher <avi@deitcher.net>

**How to categorize this PR?**

/area dev-productivity
/kind bug

**What this PR does / why we need it**:

inlets OSS has been deprecated and removed, making `hack/hook-me.sh` not work.

This replaces it with https://github.com/mvladev/quic-reverse-http-tunnel as suggested by @timebertt 

Also does 2 more things:

* replaces busybox with alpine for the debug container (so much easier to add things)
* replaces the pod with a Deployment, so it restarts if it crashes

**Special notes for your reviewer**:

This does not 100% work yet, but it is ready for eyes-on. When I run it, I get the following errors:

```
[Step 9] Running quic client
I0716 14:59:27.964727       1 main.go:81] "client started"
2021/07/16 14:59:27 Failed to sufficiently increase receive buffer size. Was: 208 kiB, wanted: 2048 kiB, got: 416 kiB.
E0716 14:59:29.971684       1 main.go:91] "could not dial quic server" err="NO_ERROR: Handshake did not complete in time"
2021/07/16 14:59:29 Failed to sufficiently increase receive buffer size. Was: 208 kiB, wanted: 2048 kiB, got: 416 kiB.
E0716 14:59:31.977065       1 main.go:91] "could not dial quic server" err="NO_ERROR: Handshake did not complete in time"
2021/07/16 14:59:31 Failed to sufficiently increase receive buffer size. Was: 208 kiB, wanted: 2048 kiB, got: 416 kiB.
```

**Release note**:

```bugdix developer
Fix hook-me.sh to work again
```

### CODE REVIEW

**Comment by gardener-robot-ci-2** [comment]
Thank you @deitch for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

**Comment by deitch** [comment]
Almost there. Apparently, the client from [the project](https://github.com/mvladev/quic-reverse-http-tunnel) is hard-coded to look for a certificate with the name `quic-tunnel-server`. Our certs obviously don't use that. So we will need to use separate certs.

**Comment by deitch** [comment]
OK, this looks rather nice now. Ready for review.

**Comment by deitch** [comment]
Something isn't 100% right. The tunnel is connected, but traffic does not go through correctly, just kind of hangs. I probably configured it wrong.

**Comment by deitch** [comment]
It looks like ghost-tunnel is doing its job - terminating TLS request from the api server and then passing it as tcp to the quic tunnel - which is passing it back to the extension as admission controller. But then it fails because admission controller expects https, not http (tcpdump output):

```
14:35:21.742864 IP (tos 0x0, ttl 64, id 16111, offset 0, flags [DF], proto TCP (6), length 1238)
    localhost.58842 > localhost.8080: Flags [P.], cksum 0x02cb (incorrect -> 0x2b7f), seq 1:1187, ack 1, win 342, options [nop,nop,TS val 3415155204 ecr 3415155203], length 1186: HTTP, length: 1186
	POST /shoot?timeout=10s HTTP/1.1
	Host: gardener-extension-provider-packet.extension-provider-packet-9r7xh:443
	User-Agent: kube-apiserver-admission
	Content-Length: 10461
	Accept: application/json, */*
	Content-Type: application/json
	Accept-Encoding: gzip

	{"kind":"AdmissionReview","apiVersion":"admission.k8s.io/v1beta1","request":{"uid":"ffc28e28-ce57-4bc2-9a25-0828ee8335c8","kind":{"group":"apps","version":"v1","kind":"Deployment"},"resource":{"group":"apps","version":"v1","resource":"deployments"},"requestKind":{"group":"apps","version":"v1","kind":"Deployment"},"requestResource":{"group":"apps","version":"v1","resource":"deployments"},"name":"vpn-shoot","namespace":"kube-system","operation":"UPDATE","userInfo":{"username":"gardener.cloud:system:gardener-resource-manager","groups":["system:masters","system:authenticated"]},"object":{"kind":"Deployment","apiVersion":"apps/v1","metadata":{"name":"vpn-shoot","namespace":"kube-system","selfLink":"/apis/apps/v1/namespaces/kube-system/deployments/vpn-shoot","uid":"490b8fd4-e20a-4ed8-bbbf-abf642e279db","resourceVersion":"7196836","generation":138,"creationTimestamp":"2021-06-07T11:34:40Z","labels":{"app":"vpn-shoot"," [|http]
14:35:21.742869 IP (tos 0x0, ttl 64, id 27622, offset 0, flags [DF], proto TCP (6), length 52)
    localhost.8080 > localhost.58842: Flags [.], cksum 0xfe28 (incorrect -> 0x88c4), seq 1, ack 1187, win 333, options     localhost.8080 > localhost.58842: Flags [P.], cksum 0xfe74 (incorrect -> 0xa867), seq 1:77, ack 10723, win 342, options [nop,nop,TS val 3415155473 ecr 3415155204], length 76: HTTP, length: 76
	HTTP/1.0 400 Bad Request

	Client sent an HTTP request to an HTTPS server.
```

Which makes sense, but then how in the world did this work before? inlets also had ghostunnel doing the same thing.

**Comment by deitch** [comment]
It works! It actually works! With quic-tunnel, since the connection is tcp on either end, I don't need the extra ghostunnel or certs, just the straight tunnel. I just got my cluster happily churning away with this.

This is ready for review.

File: hack/hook-me.sh [hunk 1]
```diff
@@ -47,184 +55,255 @@ spec:
...
 EOF
 }
 
-waitForInletsPodToBeReady(){
+waitForQuicDeployToBeReady(){
     namespace=${1:-}
     [[ -z $namespace ]] && echo "Please specify extension namespace!" && exit 1
 
-    until test "$(kubectl -n $namespace get pods inlets-server --no-headers | awk '{print $2}')" = "3/3"
+    until test "$(kubectl -n $namespace get deploy quic-server --no-headers | awk '{print $2}')" = "1/1"
```
Comment (dkistner) [inline hack/hook-me.sh:207]: Should we not rather check if the respective pod of the deployment is ready?

**Review by dkistner — COMMENTED** [review]
Implementation wise this looks good to me.
Just a small comment.

I cannot comment if using @mvladev [quick tunnel](https://github.com/mvladev/quic-reverse-http-tunnel) is the right choice to replace the inlets. I know this is just used for dev purposes therefore the criticality is low and requirements regarding maintenance are also not that critical as for components used in production.
@mvladev Could you please comment on this and also have another look?

/unassign
/needs second-opinion

**Review by mvladev — APPROVED** [review]
/lgtm

**Pull request merged**